src/future/shared.rs

```diff
@@ -0,0 +1,222 @@
+                               -> Result<Async<SharedItem<F::Item>>, SharedError<F::Error>> {
+        match result {
+            Ok(item) => Ok(Async::Ready(item)),
+            Err(error) => Err(error),
+        }
+    }
+
+    fn unpark_tasks(&self, tasks_receiver: &mut Receiver<Task>) {
+        self.inner.tasks_unpark_started.store(true, Ordering::Relaxed);
+        loop {
+            match tasks_receiver.try_recv() {
+                Ok(task) => task.unpark(),
+                _ => break,
+            }
+        }
+    }
+}
+
+pub fn new<F>(future: F) -> Shared<F>
+    where F: Future
+{
+    let (tasks_sender, tasks_receiver) = channel();
+    Shared {
+        inner: Arc::new(Inner {
+            synced_inner: Lock::new(SyncedInner {
+                original_future: future,
+                tasks_receiver: tasks_receiver,
+            }),
+            tasks_unpark_started: AtomicBool::new(false),
+            result: RwLock::new(None),
+        }),
+        tasks_sender: tasks_sender,
+    }
+}
+
+impl<F> Future for Shared<F>
+    where F: Future
+{
+    type Item = SharedItem<F::Item>;
+    type Error = SharedError<F::Error>;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // The logic is as follows:
+        // 1. Check if the result is ready (with tasks_unpark_started)
+        //  - If the result is ready, return it.
+        //  - Otherwise:
+        // 2. Try lock the self.inner.original_future:
+        //    - If successfully locked, poll the original future.
+        //      If the future is ready, unpark the tasks from
+        //      self.inner.tasks_receiver and return the result.
+        //    - If the future is not ready:
+        // 3. Create a task and send it through self.tasks_sender.
+        // 4. Check again if the result is ready (with tasks_unpark_started)
+        // 5. Return the result if it's ready. It is necessary because otherwise there could be
+        //    a race between the task sending and the thread receiving the tasks.
+
+        // If the result is ready, just return it
+        if self.inner.tasks_unpark_started.load(Ordering::Relaxed) {
+            return Self::result_to_polled_result(self.inner
+                .result
+                .read()
+                .unwrap()
+                .clone()
+                .unwrap());
+        }
+
+        // The result was not ready.
+        // Try lock the original future.
+        match self.inner.synced_inner.try_lock() {
+            Some(mut synced_inner) => {
+                let ref mut synced_inner = *synced_inner;
+                let ref mut original_future = synced_inner.original_future;
+                let mut result = self.inner.result.write().unwrap();
+                // Other thread could already poll the result, so we check if result has a value
+                if result.is_none() {
+                    match original_future.poll() {
```

Comment:
No thread will try to lock for read when poll is called because tasks_unpark_started is false. 
If it's true and a thread locks for write, it won't call poll() because result is not None.